File: solutions/c/binary-search/1/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length){
    size_t left = 0;
    size_t right = length - 1;
    if (!length || value < arr[left] || value > arr[right])
        return NULL;
    else 
        while (left<= right) {
            size_t mid = (right+left)/2; 
            int v = arr[mid];
            if(v == value ) 
                return (const int *)(arr + mid);
            else if (v >value) 
                right = mid - 1;
            else 
                left = mid +1;           
            }
    return NULL;
}

File: solutions/fortran/binary-search/1/binary_search.f90
module binary_search
  implicit none
contains

  function find(array, val) result(idx)
    integer, dimension(:), intent(in) :: array
    integer, intent(in) :: val
    integer :: idx
    integer :: left 
    integer :: right 
    integer :: mid 

    left = 1
    right = size(array)
    idx = -1
    if (right>0) then
     do while (left<=right)
          mid = (right+left)/2
          if (array(mid) == val) then
            idx = mid  
            right = 0
          else if (array(mid)<val) then
            left = (mid + 1)
          else
            right = (mid - 1) 
          end if
      end do
    end if
    
     
  end function

end module
